source/code/plugin/out_mdm.rb

```diff
@@ -0,0 +1,219 @@
+      aks_region = ENV['AKS_REGION']
+      if aks_resource_id.to_s.empty?
+        @log.info "Environment Variable AKS_RESOURCE_ID is not set.. "
+        raise Exception.new "Environment Variable AKS_RESOURCE_ID is not set!!" 
+      end
+      if aks_region.to_s.empty?
+        @log.info "Environment Variable AKS_REGION is not set.. "
+        raise Exception.new "Environment Variable AKS_REGION is not set!!" 
+      end
+      @@post_request_url = @@post_request_url_template % {aks_region: aks_region, aks_resource_id: aks_resource_id}
+      @log.info "POST Request url: #{@@post_request_url}"
+      ApplicationInsightsUtility.sendCustomEvent("AKSCustomMetricsMDMPluginStart", {})
+    end
+
+    # get the access token only if the time to expiry is less than 5 minutes
+    def get_access_token
+      if @cached_access_token.to_s.empty? || (Time.now + 5*60 > @token_expiry_time) # token is valid for 60 minutes. Refresh token 5 minutes from expiration
+        @log.info "Refreshing access token for out_mdm plugin.."
+        token_uri = URI.parse(@token_url)
+        http_access_token = Net::HTTP.new(token_uri.host, token_uri.port)
+        http_access_token.use_ssl = true
+        token_request = Net::HTTP::Post.new(token_uri.request_uri)
+        token_request.set_form_data(
+          {
+            'grant_type' => @@grant_type, 
+            'client_id' => @data_hash['aadClientId'], 
+            'client_secret' => @data_hash['aadClientSecret'],
+            'resource' => @@token_resource_url
+            }
+        )
+        
+        token_response = http_access_token.request(token_request)
+        # Handle the case where the response is not 200 
+        parsed_json = JSON.parse(token_response.body)
+        @token_expiry_time = Time.now + 59*60 # set the expiry time to be ~one hour from current time
+        @cached_access_token = parsed_json['access_token']
+      end
+      @cached_access_token
+    end 
+
+    def write_status_file(success, message)
+      fn = '/var/opt/microsoft/omsagent/log/MDMIngestion.status'
+      status = '{ "operation": "MDMIngestion", "success": "%s", "message": "%s" }' % [success, message]
+      begin
+        File.open(fn,'w') { |file| file.write(status) }
+      rescue => e
+        @log.debug "Error:'#{e}'"
+        ApplicationInsightsUtility.sendExceptionTelemetry(e.backtrace)
+      end
+    end
+
+    # This method is called when an event reaches to Fluentd.
+    # Convert the event to a raw string.
+    def format(tag, time, record)
+      if record != {}
+        @log.trace "Buffering #{tag}"
+        return [tag, record].to_msgpack
+      else
+        return ""
+      end
+    end
+
+    # This method is called every flush interval. Send the buffer chunk to MDM. 
+    # 'chunk' is a buffer chunk that includes multiple formatted records
+    def write(chunk)
+      if !@first_post_attempt_made || (Time.now > @last_post_attempt_time + retry_mdm_post_wait_minutes*60)
```

Comment:
Can you please explain what you are trying to do here ? I see you set both first_ & last_ only when you see 403. So initially you see 200 and then if you start to see 403, whats the behavior ?